MapboxCoreNavigation/Offline.swift

```diff
@@ -0,0 +1,122 @@
+import Foundation
+import MapboxDirections
+import MapboxNavigationNative
+
+public enum RequestOption {
+    case preferServerSide // TODO: Not yet implemented
+    case preferClientSide // TODO: Not yet implemented
+    case serverSideOnly
+    case clientSideOnly
+}
+
+public enum OfflineRoutingError: Error {
+    case unexpectedRouteResult
+    case corruptRouteData
+}
+
+class OfflineStorage: NSObject {
+    
+    static var tilesPath: URL {
+        let path = coreNavigationCacheDirectory.appendingPathComponent("tiles")
+        ensureDirectoryExist(at: path)
+        return path
+    }
+    
+    static var translationsPath: URL {
+        let path = coreNavigationCacheDirectory.appendingPathComponent("translations")
+        ensureDirectoryExist(at: path)
+        return path
+    }
+    
+    static var coreNavigationCacheDirectory: URL {
+        let cacheDirectory = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)[0]
+        return URL(fileURLWithPath: cacheDirectory).appendingPathComponent(Bundle.mapboxCoreNavigation.bundleIdentifier!).appendingPathComponent("offline")
+    }
+    
+    // Ensures that a directory exist at the given path.
+    // Returns false if no directory was created.
+    @discardableResult
+    static func ensureDirectoryExist(at path: URL) -> Bool {
+        do {
+            try FileManager.default.createDirectory(at: path, withIntermediateDirectories: true, attributes: nil)
+            return true
+        } catch {
+            return false
+        }
+    }
+}
+
+public typealias RouteCompletionHandler = (_ waypoints: [Waypoint]?, _ routes: [Route]?, _ error: NSError?) -> Void
+
+public class ExtendedDirections: NSObject {
+    
+    public static let shared = ExtendedDirections()
+    
+    static let offlineSerialQueueLabel = Bundle.mapboxCoreNavigation.bundleIdentifier!.appending(".offline")
+    
+    static let serialQueue = DispatchQueue(label: ExtendedDirections.offlineSerialQueueLabel)
+    
+    var _navigator: MBNavigator!
+    var navigator: MBNavigator {
+        //assert(OperationQueue.current?.underlyingQueue?.label == ExtendedDirections.offlineSerialQueueLabel,
+        //       "The offline navigator must be accessed from the dedicated serial queue")
+        if _navigator == nil {
+            self._navigator = MBNavigator()
+            
+            let tilePath = OfflineStorage.tilesPath.absoluteString.replacingOccurrences(of: "file://", with: "")
+            let localizationPath = OfflineStorage.translationsPath.absoluteString.replacingOccurrences(of: "file://", with: "")
+            
+            // navigator raises an exception after each try:
+            // 1: uncompressed tiles packed in a tar archive,
+            // 2: uncompressed tiles in directories
+            // 3: compressed tiles in directories
+            // We are using the 3rd option
```

Comment:
Should we capture these exceptions? If so, I think we would have to capture them in Obj-C.

Reply:
If that's the case, do we need the `OfflineDirections` class to be written in ObjC? Or another intermediary wrapper perhaps? Obviously we can't have these exceptions (which seem like they could be used for control flow or cascading workflow options) to crash the containing app.